### PR TITLE
Add More Coil and Tank Choices for Desuperheater

### PR DESCRIPTION
**Pull request by joseph-robertson**
Pull request overview
---------------------

Ability to have on `CoilWaterHeatingDesuperheater`:
* `WaterHeaterStratified` tank object
* `CoilCoolingWaterToAirHeatPumpEquationFit` and `CoilCoolingDXMultiSpeed` heating source types

https://github.com/NREL/EnergyPlus/pull/7251
https://github.com/NREL/OpenStudio/pull/3565

This Pull Request is concerning:

 - [x] **Case 1 - `NewTest`:** a new test for a new model API class,
 - [ ] **Case 2 - `TestFix`:** a fix for an existing test. The GitHub issue should be referenced in the PR description
 - [x] **Case 3 - `NewTestForExisting`:** a new test for an already-existing model API class
 - [ ] **Case 4 - `Other`:** Something else, like maintenance of the repo, or just committing test results with a new OpenStudio version.

**This is a bit of a mix between case 1 and 3, because the class was existing, but the test needs a recent PR to run as it added new coil and heat rejection target types.**

----------------------------------------------------------------------------------------------------------

### Case 1/3: New test for an extisting model API class but with recent features

This pull request adds missing tests for the following classes but also relies on a new PR:
*  [CoilWaterHeatingDesuperheater](https://github.com/NREL/OpenStudio/blob/develop/openstudiocore/src/model/CoilWaterHeatingDesuperheater.hpp)
* In order for the test to be able to run, https://github.com/NREL/OpenStudio/pull/3565 is required. Which means the OSM should be added once 2.9.0 is out

It will test two cases:
* A `WaterHeaterStratified` as a Heat Rejection Target, and a Coil = `CoilCoolingWaterToAirHeatPumpEquationFit`
* A `WaterHeaterMixed` as a Heat Rejection Target, and a Coil = `CoilCoolingDXMultiSpeed`

#### Work Checklist

The following has been checked to ensure compliance with the guidelines:

 - [ ] Tests pass either:

     - [ ] with current develop (incude SHA): **NO, this needs https://github.com/NREL/OpenStudio/pull/3565, so add it after 2.9.0 is out**
         - [x] The label `PendingOSM` has been added to this PR
         - [x] A matching OSM test has not yet been added because the official release is pending, but `model_tests.rb` has a TODO.
            ```ruby
            def test_airterminal_cooledbeam_rb
              result = sim_test('airterminal_cooledbeam.rb')
            end

            # TODO : To be added once the next official release
            # including this object is out : 2.5.0
            # def test_airterminal_fourpipebeam_osm
            #   result = sim_test('airterminal_fourpipebeam.osm')
            # end
            ```
        - [ ] No `out.osw` have been committed as they need to be run with an official OpenStudio version

 - [ ] **Ruby test is stable**: when run multiple times on the same machine, it produces the same total site kBTU.
     - [x] I ensured that I assign systems/loads/etc in a repeatable manner (eg: if I assign stuff to thermalZones, I do `model.getThermalZones.sort_by{|z| z.name.to_s}.each do ...` so I am sure I put the same ZoneHVAC systems to the same zones regardless of their order)
     - [ ] I tested stability using `process_results.py` (see `python process_results.py --help` for usage).
     Please paste the heatmap png generated after running the following commands:
        ```bash
        # Clean up all custom-tagged OSWs
        python process_results.py test-stability clean
        # Run your test 5 times in a row. Replace `testname_rb` (eg `airterminal_fourpipebeam_rb`)
        python process_results.py test-stability run -n testname_rb
        # Check that they all passed
        python process_results.py test-status --tagged
        # Check site kBTU differences
        python process_results.py heatmap --tagged

        ```

----------------------------------------------------------------------------------------------------------

### Review Checklist

 - [ ] Code style (indentation, variable names, strip trailing spaces)
 - [ ] Functional code review (it has to work!)
 - [ ] Matching OSM test has been added or `# TODO` added to `model_tests.rb`
 - [ ] Appropriate `out.osw` have been committed
 - [ ] Test is stable
 - [ ] The appropriate labels have been added to this PR:
   - [ ] One of: `NewTest`, `TestFix`, `NewTestForExisting`, `Other`
   - [ ] If `NewTest`: add `PendingOSM` if needed

### CODE REVIEW

File: model/simulationtests/gshp_desuperheater.rb [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+model.add_design_days()
+
+#create schedule object
+schedule_constant = OpenStudio::Model::ScheduleConstant.new(model)
+
+#create desuperheater object
+coil_water_heating_desuperheater = OpenStudio::Model::CoilWaterHeatingDesuperheater.new(model, schedule_constant)
+
+#create stratified water heater
+water_heater_stratified = OpenStudio::Model::WaterHeaterStratified.new(model)
```
Comment (shorowit) [inline model/simulationtests/gshp_desuperheater.rb]: Great, you're testing both GSHP coil _and_ stratified water heater. 👍

Comment (shorowit) [reply model/simulationtests/gshp_desuperheater.rb]: Can you add a test for multispeed coil too?

Comment (joseph-robertson) [reply model/simulationtests/gshp_desuperheater.rb]: Yes. And I'm going to rename this test to `test_coil_waterheating_desuperheater_rb`.

**Comment by jmarrec** [comment]
@joseph-robertson It'd be great if you could actually use the PR template when you open a PR. I know it looks intimidating (and annoyingly long) but it's not really, and does include a lot of checks that are useful to see if the PR is ready to drop in.

**Comment by jmarrec** [comment]
@joseph-robertson I revamped the entire test to produce valid loops.

E+ is failing though:
```
   ** Severe  ** Coil:WaterHeating:Desuperheater = COIL WATER HEATING DESUPERHEATER 1: Rated Heat Reclaim Recovery Efficiency must be > 0.0 and <= 0.3, Rated Heat Reclaim Recovery Efficiency = 0.800
   ** Severe  ** Coil:WaterHeating:Desuperheater = COIL WATER HEATING DESUPERHEATER 2: Rated Heat Reclaim Recovery Efficiency must be > 0.0 and <= 0.3, Rated Heat Reclaim Recovery Efficiency = 0.800
   **  Fatal  ** Errors found in getting Coil:WaterHeating:Desuperheater input. Preceding condition causes termination.
   ...Summary of Errors that led to program termination:
   ..... Reference severe error count=2
   ..... Last severe error=Coil:WaterHeating:Desuperheater = COIL WATER HEATING DESUPERHEATER 2: Rated Heat Reclaim Recovery Efficiency must be > 0.0 and <= 0.3, Rated Heat Reclaim Recovery Efficiency = 0.800
```

From E+ idd:
```
  N2 , \field Rated Heat Reclaim Recovery Efficiency
       \type real
       \minimum> 0.0
       \note Enter the fraction of waste heat reclaimed by the desuperheater water heating coil.
```

Seems like that guy should have a `\maximum< 0.3` or something.

Anyways, it means we should probably handle that on OpenStudio's side, where we currently set it to 0.8 (for some reason)

**Comment by joseph-robertson** [comment]
I'll use the template in the future.

**Pull request merged**